consensus/src/checkpoint.rs

```diff
@@ -112,47 +117,221 @@ mod tests {
+        let root_block_height =
+            gen_num_blocks_in_epochs(slot_fill_rate, slot_fill_rate_delta, root_epoch_int as f64);
+        return (root_epoch_int, root_block_height as u32);
+    }
+
+    fn convert(x: f64) -> i32 {
+        x.round().rem_euclid(2f64.powi(32)) as u32 as i32
+    }
+
+    fn gen_num_blocks_in_slots(slot_fill_rate: f64, slot_fill_rate_delta: f64, n: f64) -> i32 {
+        let min_blocks = n * f64::max(slot_fill_rate - slot_fill_rate_delta, 0.0);
+        let max_blocks = n * f64::min(slot_fill_rate + slot_fill_rate_delta, 1.0);
+        let num_blocks_in_slots = thread_rng().gen_range(min_blocks..max_blocks);
+        let num = convert(num_blocks_in_slots);
+        return num;
+    }
+
+    fn gen_spot(block: &mut ProtocolState) {
+        // New default consensus state
+        // Generate blockchain position and epoch lengths.
+        // staking_epoch == root_epoch, next_staking_epoch == root_epoch + 1
+        // New state slot and length.
+        // New total currency for state.
+        // New epoch data for staking and next epochs.
+        // New block reward information (unused in chain selection).
+        let consensus_state = ConsensusState::new();
+        block.body.consensus_state = consensus_state;
+    }
+
+    fn gen_spot_pair_common_checkpoints(
+        a: &mut ProtocolState,
+        b: &mut ProtocolState,
+        min_a_curr_epoch_slot: u32,
+    ) {
+        // New pairs of spot blocks that share common checkpoints.
+        // The overlap of the checkpoints and the root epoch positions of the blocks
+        // that are generated can be configured independently so that this function
+        // can be used in other generators that wish to generates pairs of spot blocks
+        // with specific constraints.
+        let default_slot_fill_rate = 0.65;
+        let default_slot_fill_rate_delta = 0.15;
+        let base_root_epoch_position =
+            gen_spot_root_epoch_position(default_slot_fill_rate, default_slot_fill_rate_delta);
+
+        // Constraining the second state to have a greater blockchain length than the
+        // first, we need to constrain the first blockchain length such that there is some room
+        // leftover in the epoch for at least 1 more block to be generated.
+        let blockchain_length_relativity = Some(&a);
+        let max_epoch_slot = match blockchain_length_relativity {
+            Some(blockchain_length_relativity) => {
+                a.body.consensus_state.curr_global_slot.slots_per_epoch.0 - 4
+            } // -1 to bring into inclusive range, -3 to provide 2 slots of fudge room
+            None => a.body.consensus_state.curr_global_slot.slots_per_epoch.0 - 1, // -1 to bring into inclusive range
+        };
+
+        let min_a_curr_epoch_slot_defaut = 0;
+        let min_a_curr_epoch_slot_sum = min_a_curr_epoch_slot_defaut + min_a_curr_epoch_slot;
+        let slot = thread_rng().gen_range(min_a_curr_epoch_slot_sum..max_epoch_slot);
```

Comment:
U can generate f64 number directly, see https://docs.rs/rand/latest/rand/trait.Rng.html#example-1

Reply:
the slot is an integer type, I use slot as f64 just for gen_num_blocks_in_slots perform calculate with float number const.